go/pkg/sysdb/coordinator/table_catalog.go

```diff
@@ -859,3 +942,149 @@ func (tc *Catalog) FlushCollectionCompaction(ctx context.Context, flushCollectio
+		existingVersion := int64(collectionEntry.Version)
+
+		// Do a check to see if the version is stale.
+		if existingVersion > versionAtCompactionStart {
+			// Compactor is trying to flush a version that is no longer valid, since
+			// a different compaction instance has already incremented the version.
+			log.Info("Compactor is trying to flush a stale version", zap.Int64("existing_version", existingVersion), zap.Int64("current_collection_version", versionAtCompactionStart))
+			return nil, common.ErrCollectionVersionStale
+		}
+
+		existingVersionFileName := collectionEntry.VersionFileName
+		// Read the VersionFile from S3MetaStore.
+		existingVersionFilePb, err := tc.s3Store.GetVersionFile(flushCollectionCompaction.TenantID, flushCollectionCompaction.ID.String(), existingVersion, existingVersionFileName)
+		if err != nil {
+			return nil, err
+		}
+
+		// The update function takes the content of the existing version file,
+		// and the set of segments that are part of the new version file.
+		// NEW VersionFile is created in S3 at this step.
+		newVersionFileName, err := tc.updateVersionFileInS3(ctx, existingVersionFilePb, flushCollectionCompaction)
+		if err != nil {
+			return nil, err
+		}
+
+		txErr := tc.txImpl.Transaction(ctx, func(txCtx context.Context) error {
+			// register files to Segment metadata
+			err = tc.metaDomain.SegmentDb(txCtx).RegisterFilePaths(flushCollectionCompaction.FlushSegmentCompactions)
+			if err != nil {
+				return err
+			}
+
+			// At this point, a concurrent Transaction can still update/commit
+			// the Collection entry.
+			// Since this Tx is ReadCommitted, the result of other Tx will be
+			// visible to the statement below. Hence the statement below will
+			// use WHERE clause to ensure that its update will not go through
+			// if the Collection entry is updated by another Tx.
+
+			// Update collection log position and version
+			rowsAffected, err := tc.metaDomain.CollectionDb(txCtx).UpdateLogPositionAndVersionInfo(
+				flushCollectionCompaction.ID.String(),
+				flushCollectionCompaction.LogPosition,
+				flushCollectionCompaction.CurrentCollectionVersion,
+				existingVersionFileName,
+				flushCollectionCompaction.CurrentCollectionVersion+1,
+				newVersionFileName,
+			)
+			if err != nil {
+				return err
+			}
+			if rowsAffected == 0 {
+				// CAS operation failed.
+				// Error out the transaction, so that segment is not updated.
+				return common.ErrCollectionEntryIsStale
+			}
+
+			// CAS operation succeeded. Update tenant compaction time and then
+			// COMMIT the transaction.
+
+			// update tenant last compaction time
+			// TODO: add a system configuration to disable
+			// since this might cause resource contention if one tenant has a lot of collection compactions at the same time
+			lastCompactionTime := time.Now().Unix()
+			err = tc.metaDomain.TenantDb(txCtx).UpdateTenantLastCompactionTime(flushCollectionCompaction.TenantID, lastCompactionTime)
```

Comment:
this does not seem right with read committed isolation level? Can the following sequence happen?

```
TX1 updates segments and collections
TX2 updates segments and collections
TX2 updates last compaction time
TX1 updates last compaction time
```

So even though the compaction and segments are of TX2, the last compaction time is of TX1?